Blame/filter in Browse, replace formfilehistory

gerhardol opened the pull request:
Closes #7598

Based on #9424.
This PR could be split in several PRs, submitted as is to describe the "initial completion" of #7598 and describe what could be improved.

I consider this to be ready for review, but may tweak the functionality until #9424 is merged.

## Proposed changes

* See screenshots below

* RevDiff and FileTree selects filtered files by default when changing commits, the filtered file is "followed" when switching commits (unless another file is selected and it exists in the next commit, as already implemented).

* Some refactoring to allow for new functionality separated.

Future:
* Present active filters better. Some way to deactivate filters in a simple way. (Some discussion in #7598).
* A few tests could be added (some added)

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

### After

Show path filter in title (last part is due to Debug build, unchanged)
![image](https://user-images.githubusercontent.com/6248932/128436478-32a5a9c5-0262-40b7-97c2-c5743f53ff8f.png)

Separator changed as T/F/H/B hotkey commands are related

![image](https://user-images.githubusercontent.com/6248932/128781499-c2b4d391-2f76-4a87-bfda-477f10eac852.png)

Context command, hotkey `F` to filter a file in the grid
(similar to open a new instance)

![image](https://user-images.githubusercontent.com/6248932/128781554-8478a4e9-5de3-40e6-8f25-7f6c9272b4d2.png)

![image](https://user-images.githubusercontent.com/6248932/128781827-f36a6214-1d24-4072-ab5c-286bd9042a99.png)

Configuration to open a Browse with filter instead of FileHistory.
Enabled by default.
This is put in Browse as Appearance is too busy already

![image](https://user-images.githubusercontent.com/6248932/128436980-ba9a5c26-f4b5-4d49-8261-b581ffc068fd.png)

## Test methodology <!-- How did you ensure quality? -->

manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
Rebased to resolve merge conflicts (also in #9424 that this PR is based on).
Add clear of title similar when advanced filter is modified.

RussKie commented on GitUI/CommandsDialogs/RevisionFileTreeControl.cs:
```diff
@@ -149,6 +150,8 @@ public void InvokeFindFileDialog()
             findToolStripMenuItem_Click(this, EventArgs.Empty);
         }
 
+        public string? FileToSelectInitially { private get; set; } = null;
```
ditto

gerhardol replied:
Handled as above

RussKie reviewed (approved):
I'll need your help retrofitting the filtering to https://github.com/gitextensions/gitextensions/pull/9580 after this is merged.

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -216,6 +220,8 @@ public void DisplayDiffTab(IReadOnlyList<GitRevision> revisions)
             }
         }
 
+        public string? FileToSelectInitially { private get; set; } = null;
```
Adding documentation, but prefer to not change the name (to `SelectedFile`, other suggestions?).  The file selection is "followed" when you select a new commit. So if you change commit with a filter, the diff tab is similar to FileHistory.
There is a similar functionality that try to follow the last selected file too `oldDiffItem`. If that file is not in next commit, then `FileToSelectInitially` is selected again.

RussKie replied:
Would `FollowedFile` work?

gerhardol replied:
FallbackFollowedFile, the selected file is followed first

gerhardol commented:
Squashed some and rebased, adding a fixup
Plan to squash the fixup and rebase merge after approval

> I'll need your help retrofitting the filtering to #9580 after this is merged.

Do you want me to rebase on this PR (could be done before this is merged, I can push to my repo).

RussKie commented:
> Do you want me to rebase on this PR (could be done before this is merged, I can push to my repo).

Let's merge this first. Then if you can help rebasing my PR on top of the latest master - that'd be grand. My fork is open, so feel free to push in to it.

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -216,6 +220,12 @@ public void DisplayDiffTab(IReadOnlyList<GitRevision> revisions)
             }
         }
 
+        /// <summary>
+        /// The file in the list to select initially (and fallback if the previous selected file
+        /// cannot be "followed" after selecting next commit).
+        /// </summary>
```
Did I get it right?
```suggestion
        /// <summary>
        ///  Gets or sets the file in the list to select initially and that is being "followed" when switching between commits.        
        ///  This is also used as a fallback, if the previously selected file cannot be "followed" after selecting a different commit.
        /// </summary>
```

gerhardol replied:
Yes, will use something like this

RussKie reviewed (commented):
👍

RussKie commented:
Merge, when ready
